clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1829,15 +2038,31 @@ class SyclKernelUnionChecker : public SyclKernelFieldHandler {
     return true;
   }
 
+  bool enterUnion(const CXXRecordDecl *, ParmVarDecl *) override {
+    ++UnionCount;
```

Comment:
The union checker now diagnoses a use of a union.

Reply:
Just an FYI - The FieldChecker visitor does this traversal already.  So the handlers in this visitor can catch it. Having said that, I do think dealing with structs in a follow-up PR makes sense because of the time constraint we have with this PR. It simplifies concerns we have with the BodyCreator at the moment.